biogps <- readRDS('data-raw/single-cell/biogps/biogps.rds')
cell_info <- readRDS('data-raw/cell_info/cell_info.rds')
# gslist.kegg <- readRDS('data-raw/padog/gslist.rds')
# gs.names.kegg <- readRDS('data-raw/padog/gs.names.rds')
# gslist.go <- readRDS('data-raw/padog/gslist.go.rds')
# gs.names.go <- readRDS('data-raw/padog/gs.names.go.rds')
genes <- readRDS('data-raw/genes/genes.rds')
pert_names <- readRDS('data-raw/drug_gene_queries/pert_names.rds')
tx2gene <- readRDS('data-raw/tx2gene/tx2gene.rds')
tx2gene_mouse <- readRDS('data-raw/tx2gene/tx2gene_mouse.rds')
homologene <- readRDS('data-raw/homologene/homologene.rds')
hs <- readRDS('data-raw/homologene/hs.rds')

usethis::use_data(biogps, cell_info, genes, pert_names, hs, homologene, tx2gene, tx2gene_mouse, internal = TRUE, overwrite = TRUE)
